File: src/IsbInterop/Accessory/Proxies/List.vb
﻿Namespace Accessory.Proxies

  ''' <summary>
  ''' Обертка над IList.
  ''' </summary>
  Friend Class List(Of TI As IBaseIsbObject)
    Inherits ForEach(Of TI)
    Implements IList(Of TI)

    ''' <summary>
    ''' Добавляет элемент в список.
    ''' </summary>
    ''' <typeparam name="TP">Тип параметра.</typeparam>
    ''' <param name="name">Имя элемента.</param>
    ''' <param name="value">Значение элемента.</param>
    Public Sub Add(Of TP As IBaseIsbObject)(name As String, value As TP) Implements IList(Of TI).Add
      If TypeOf value Is TI Then
        InvokeRcwInstanceMethod("Add", New Object() {name, DirectCast(value, IRcwProxy).RcwObject})
      Else
        Throw New InvalidOperationException(String.Format("Cannot convert value to {0}", GetType(TI)))
      End If
    End Sub

    ''' <summary>
    ''' Добавляет элемент в список.
    ''' </summary>
    ''' <param name="name">Имя элемента.</param>
    ''' <param name="value">Значение элемента.</param>
    Public Sub Add(name As String, value As Object) Implements IList(Of TI).Add
      InvokeRcwInstanceMethod("Add", New Object() {name, value})
    End Sub

    ''' <summary>
    ''' Получает значение элемента списка по индексу.
    ''' </summary>
    ''' <param name="index">Индекс элемента в списке.</param>
    ''' <returns>Значение элемента с указанным индексом.</returns>
    Public Function GetValues(index As Integer) As TI Implements IList(Of TI).GetValues
      Dim rcwObjectResult = GetRcwProperty("Values", index)
      Return IsbObjectResolver.Resolve(Of TI)(rcwObjectResult, Scope)
    End Function

    ''' <summary>
    ''' Получает значение элемента по имени.
    ''' </summary>
    ''' <param name="name">Имя элемента.</param>
    ''' <returns>Значение элемента.</returns>
    Public Function GetValueByName(name As String) As TI Implements IList(Of TI).GetValueByName
      Dim rcwObject = InvokeRcwInstanceMethod("ValueByName", New Object() {name})
      Return IsbObjectResolver.Resolve(Of TI)(rcwObject, Scope)
    End Function

    ''' <summary>
    ''' Установливает значение элемента.
    ''' </summary>
    ''' <param name="name">Имя элемента.</param>
    ''' <param name="value">Значение элемента.</param>
    Public Sub SetVar(name As String, value As Object) Implements IList(Of TI).SetVar
      Dim isbObject = TryCast(value, IRcwProxy)

      If isbObject IsNot Nothing Then
        InvokeRcwInstanceMethod("SetVar", New Object() {name, isbObject.RcwObject})
      Else
        InvokeRcwInstanceMethod("SetVar", New Object() {name, value})
      End If
    End Sub

    ''' <summary>
    ''' Конструктор.
    ''' </summary>
    ''' <param name="rcwIList">COM-объект IList.</param>
    ''' <param name="scope">Область видимости.</param>
    Public Sub New(rcwIList As Object, scope As IScope)
      MyBase.New(rcwIList, scope)
    End Sub
  End Class
End Namespace

File: src/IsbInterop.CSharp/Accessory/Proxy/List.cs
using System;

namespace IsbInterop.Accessory.Proxy
{
  /// <summary>
  /// Обертка над IList.
  /// </summary>
  internal class List<TI> : ForEach<TI>, IList<TI>
    where TI : IBaseIsbObject
  {
    /// <summary>
    /// Добавляет элемент в список.
    /// </summary>
    /// <typeparam name="TP">Тип параметра.</typeparam>
    /// <param name="name">Имя элемента.</param>
    /// <param name="value">Значение элемента.</param>
    public void Add<TP>(string name, TP value) where TP : IBaseIsbObject
    {
      if (value is TI)
        InvokeRcwInstanceMethod("Add", new [] { name, ((IRcwProxy)value).RcwObject });
      else
        throw new InvalidOperationException($"Cannot convert value to {typeof(TI)}");
    }

    /// <summary>
    /// Добавляет элемент в список.
    /// </summary>
    /// <param name="name">Имя элемента.</param>
    /// <param name="value">Значение элемента.</param>
    public void Add(string name, object value)
    {
      InvokeRcwInstanceMethod("Add", new [] { name, value });
    }

    /// <summary>
    /// Получает значение элемента списка по индексу.
    /// </summary>
    /// <param name="index">Индекс элемента в списке.</param>
    /// <returns>Значение элемента с указанным индексом.</returns>
    public TI GetValues(int index)
    {
      var rcwObject = GetRcwProperty("Values", index);
      return IsbObjectResolver.Resolve<TI>(rcwObject, Scope);
    }

    /// <summary>
    /// Получает значение по имени.
    /// </summary>
    /// <param name="name">Имя искомого элемента.</param>
    /// <returns>Имя искомого элемента.</returns>
    public TI GetValueByName(string name)
    {
      var rcwObject = InvokeRcwInstanceMethod("ValueByName", new object[] { name });
      var value = IsbObjectResolver.Resolve<TI>(rcwObject, Scope);

      return value;
    }

    /// <summary>
    /// Установливает значение элемента.
    /// </summary>
    /// <param name="name">Имя элемента.</param>
    /// <param name="value">Значение элемента.</param>
    public void SetVar(string name, object value)
    {
      var isbObject = value as IRcwProxy;

      if (isbObject != null)
        InvokeRcwInstanceMethod("SetVar", new [] { name, isbObject.RcwObject });
      else
        InvokeRcwInstanceMethod("SetVar", new [] { name, value });
    }

    /// <summary>
    /// Конструктор.
    /// </summary>
    /// <param name="rcwIList">COM-объект IList.</param>
    /// <param name="scope">Область видимости.</param>
    public List(object rcwIList, IScope scope) : base(rcwIList, scope) { }
  }
}
